src/thor/bidirectional_astar.cc

```diff
@@ -916,10 +918,167 @@ void BidirectionalAStar::SetDestination(GraphReader& graphreader, const valhalla
+ * @param  costing      Costing function
+ * @param  transition_cost  Transition cost for the shortcut edge
+ * @param  previous_cost    Total cost from the previous edge
+ * TODO: add time information
+ */
+std::vector<PathInfo> recost_shortcut_forward(GraphReader& graphreader,
+                                              const GraphId& shortcut_id,
+                                              const DynamicCost& costing,
+                                              Cost transition_cost,
+                                              Cost previous_cost) {
+  auto edges = graphreader.RecoverShortcut(shortcut_id);
+
+  std::vector<PathInfo> path;
+  // transition cost will not be included for the first edge;
+  // so, add it here
+  previous_cost += transition_cost;
+
+  const auto on_new_label = [&path, &previous_cost, &transition_cost](const EdgeLabel& label) {
+    Cost transition = path.empty() ? transition_cost : label.transition_cost();
+    path.emplace_back(label.mode(), previous_cost + label.cost(), label.edgeid(), 0,
+                      label.restriction_idx(), transition);
+  };
+
+  auto edge_iter = edges.begin();
+  const auto get_next_edge = [&edge_iter, &edges]() -> GraphId {
+    if (edge_iter == edges.end())
+      return {};
+
+    return *edge_iter++;
+  };
+
+  try {
+    sif::recost_forward(graphreader, costing, get_next_edge, on_new_label);
+  } catch (const std::exception& e) {
+    LOG_ERROR(std::string("Failed to recost shortcut edge: ") + e.what());
+    return {};
+  } catch (...) {
+    LOG_ERROR("Failed to recost shortcut edge: unknown exception");
+    return {};
+  }
+
+  return path;
+}
+
+void BidirectionalAStar::FormPathForward(baldr::GraphReader& graphreader,
+                                         uint32_t last_label_idx,
+                                         std::vector<PathInfo>& path) {
+  // we should go from the beginning to the end because of possible recostings;
+  // so, at first we should collect edge label indexes
+  std::stack<uint32_t> forward_idxs;
+  // Recover forward labels using predecessor indexes
+  for (auto edgelabel_index = last_label_idx; edgelabel_index != kInvalidLabel;
+       edgelabel_index = edgelabels_forward_[edgelabel_index].predecessor()) {
+    forward_idxs.push(edgelabel_index);
+  }
+
+  // keep previous edgelabel index to
+  uint32_t predidx = kInvalidLabel;
+  // form path from forward labels
+  while (!forward_idxs.empty()) {
+    uint32_t edgelabel_idx = forward_idxs.top();
+    forward_idxs.pop();
+    const BDEdgeLabel& edgelabel = edgelabels_forward_[edgelabel_idx];
+    // get current cost
+    Cost cost = path.empty() ? Cost() : path.back().elapsed_cost;
+
+    std::vector<PathInfo> recosted_segment;
+    // recost shortcut edge
+    if (edgelabel.shortcut()) {
+      recosted_segment = recost_shortcut_forward(graphreader, edgelabel.edgeid(), *costing_,
+                                                 edgelabel.transition_cost(), cost);
```

Comment:
the last PathId stores boths the cost so far and the transition cost at the beginning of the PathID. so i think you might have the transition cost in the wrong place. i always confuse mysefl on this but thats whta it looks like to me here

Reply:
hm, didn't get this note. could you run the details ?